### DISCUSSION TITLE
Should routeTree.gen.ts be gitignored?

### DISCUSSION THREAD

**Discussion by psnehanshu** [TanStack/router#1218] [Q&A] [answered]
I belive it should be, but I want to hear what other people are doing.

**Comment by SeanCassiere** [comment]
If you are using router-cli you could probably `gitignore` the file and generate it priot to a build.

This changes however if you are using the vite-plugin, since the default vite-build script has you run `tsc` before the build. In this scenario, your build would never get past the `tsc` check.

IMHO: I just check-in the file (routeTree.gen.ts) in git, because I can't be bothered solving the vite-build script.

**Reply by gual** [reply]
I used the [Router CLI](https://tanstack.com/router/latest/docs/framework/react/guide/file-based-routing#router-cli), created a
```"router:generate": "tsr generate",``` in the `package.json` so I can use it in my build process.

**Reply by mrlubos** [reply]
@SeanCassiere has anything changed since the last year? Still the same guidance?

**Reply by SeanCassiere** [reply, marked as answer]
I've honestly gotten into the mindset of just saving your `route-tree` file in git. Purely, because the file is essential to your application running and having the typecheck phase pass correctly.

---

Updated Answer: Don't gitignore your `routeTree.gen.ts` file. This file is part of your application. It's simply managed by TanStack Router.

**Reply by SeanCassiere** [reply]
I'll probably create a FAQ section and chuck this in there.

**Reply by mrlubos** [reply]
FAQ would be nice! A counterpoint I can offer is that dependencies are also essential to your application running and would fail typecheck if not installed, yet we don't keep them in git. Even more similar, I work on codegen (see my profile) and the same question arises there. I went from yes to no on the same question. It's not that difficult to generate everything that's needed on `postinstall` or with a Vite plugin – you're doing that already anyway! In fact, I looked at the TanStack Router Vite plugin when working on mine, so I suspect it might just work™. A practical argument I hear for not committing the generated files is it makes switching between branches easier as there won't be any conflicts.

**Reply by SeanCassiere** [reply]
The FAQ is up - https://tanstack.com/router/latest/docs/framework/react/faq

@mrlubos, I get what you mean. My only argument against the conflict point is, that this isn't like your `package-lock` where some trailing commas pretty much always guarantee a conflict. So conflicts really should be at the minimum. The route tree also serves as an accurate point-in-time snapshot as to what TanStack Router would've used at runtime.

**Reply by SneakySensei** [reply]
@SeanCassiere please update the same over here as well - https://tanstack.com/start/latest/docs/framework/react/guide/routing#:~:text=You%20may%20gitignore%20this%20file%2C%20since%20it%20is%20a%20build%20artifact.

**Reply by andreilgeorgescu** [reply]
I disagree with this mindset. In my opinion, generated files should rarely ever be stored. It's like storing the user's birthdate and age instead of just the birthdate and then deriving their age from it. You end up storing more than needed while now having information that can always be incorrect or out of date (assuming the user could update their birthdate due to making a mistake).

Another example could be storing the area along with the length and width of a rectangle. The area is not only a nuisance that you must now make sure is always up to date if the length and width are ever updated, but it's even worse if it was provided to you. Because you can never be sure it's correct without essentially computing the area yourself and comparing, so there isn't even a delegation benefit. Even with no risk or storage concerns, you get pretty much nothing out of it in the large majority of situations, just more noise to sift through and mentally ignore and I see generated files as no different.

Derived state/generated files are only really worth saving if generating them is too costly or the generator isn't deterministic, neither of which seem to be the case here. CI is the place for dealing with generation, not your **source** code.

**Reply by SeanCassiere** [reply]
> I disagree with this mindset. ... by @andreilgeorgescu 

I generally I agree with this sentiment.

In this case, we've found that the `routeTree` file being ignored causes more problems given that transitive dependency upgrades can cause the output to vary slightly. Which leads in to this:

> The route tree also serves as an accurate point-in-time snapshot as to what TanStack Router would've used at runtime.

Now of course, this can be locked down, by doing a frozen lockfile install, but that is not something everyone does.

Not to mention, people not being used to the flow of `generation -> type-checking -> build` (over just `type-check -> build`) given the landscape of code generation tools not being as popular these days (unlike during the hay-day of graphql introspection).

Our stance on this, for this particular file, comes from our experience of user feedback since during the time of Router's alpha development period. To us, the `routeTree` is considered part of your _source code_ as it'd be what you'd define to have if you were to use the code-based approach and how it plays into the "type-safe" part of being a "type-safe router".

All that being said, feel free to `gitignore` it if it suits your project. I do see benefit in doing so when your route tree becomes truly to be gigantic. Our stance here, shouldn't cause any technical-restrictions which prevent you from following your project's conventions (and as far as I can see, it does not).

Happy coding 🚀

**Reply by ftzi** [reply]
`Purely, because the file is essential to your application running and having the typecheck phase pass correctly.`

If the file is deterministic and if I can generate it on demand it doesn't really make sense to commit it.

**Reply by SeanCassiere** [reply]
@ftzi that's great! Choose what best fits your project 👍🏼

**Comment by gual** [comment]
I'm using GitHub Actions so I just had to add:
```
- name: Generate routes file
      run: |
        docker compose run --rm --no-deps react npm run router:generate
```

**Comment by jessetabak** [comment]
tsc is just a compile check so I moved it after the build:

```json
"build": "vite build && tsc"
```

**Reply by psnehanshu** [reply]
What's the point of building if typecheck failed? It should ideally be 

```json
"build": "tsc --noEmit && vite build"
```

**Reply by jessetabak** [reply]
The point is that tsc will always fail if there's no generated route-tree file, which I have git ignored.
